### PR TITLE
Merge Linux v5.4.62

### PR DESCRIPTION
**Pull request by prp**
Companion PR to https://github.com/lsds/sgx-lkl/pull/822

### CODE REVIEW

**Comment by bodzhang** [comment]
This is a huge patch, 5000+ file changes. I understand this is supposed to be Linux kernel 5.4 vs 5.3 diff and the majority of the patch should be orthogonal to SGX-LKL. But it's hard to do a code review. Is it possible to break the patch to two parts, one for the pure Linux kernel diff and another for LKL/SGX-LKL changes?

**Comment by prp** [comment]
> This is a huge patch, 5000+ file changes. I understand this is supposed to be Linux kernel 5.4 vs 5.3 diff and the majority of the patch should be orthogonal to SGX-LKL. But it's hard to do a code review. Is it possible to break the patch to two parts, one for the pure Linux kernel diff and another for LKL/SGX-LKL changes?

@bodzhang It will be easiest to look at the individual commits: https://github.com/lsds/lkl/pull/10/commits/6a7ab20b67194e084ff2c48e1213ed595380af36 merges the upstream Linux v5.4 tag into our LKL repo. All merges were done automatically, except for the five conflicts listed in the commit message. These conflicts are resolved as part of https://github.com/lsds/lkl/pull/10/commits/e04e0a762e9f5de3c3594394501140f45947ca56, https://github.com/lsds/lkl/pull/10/commits/de0a1d9d802eb683efe2528dcfd0b69e1f76516d and https://github.com/lsds/lkl/pull/10/commits/8185e925ff76e285948b2c12609284f866b322f4.

(The above approach is the same how upstream LKL merged Linux 5.3 before. If you are curious about all changes between upstream Linux 5.3 and LKL, you can have a look here: https://github.com/torvalds/linux/compare/v5.3...lkl:0470fb7f23c39a7ee75aeae7fdd03f1cb2e2c064. It is mostly restricted to files under `arch/lkl` and `tools/lkl`, but there are also some changes outside of those directories.)

The SGX-LKL changes are in a separate PR https://github.com/lsds/sgx-lkl/pull/822/files, and the only meaningful change here is that the mount options for `/dev/shm` were no longer valid with Linux 5.4.

File: fs/readdir.c [hunk 1]
```diff
@@ -327,7 +327,7 @@ static int filldir64(struct dir_context *ctx, const char *name, int namlen,
 	 * Note! This range-checks 'previous' (which may be NULL).
 	 * The real range was checked in getdents
 	 */
-	if (!user_access_begin(dirent, sizeof(*dirent)))
+	if (dirent && !user_access_begin(dirent, sizeof(*dirent)))
```
Comment (davidchisnall) [inline fs/readdir.c]: This is the one I hit but didn't manage to root cause.  Why does our `user_access_begin` return success for NULL?  This feels like an LKL bug.

**Comment by davidchisnall** [comment]
Is this updating to 5.4.0 or to the latest 5.4 patch release?

**Comment by prp** [comment]
> Is this updating to 5.4.0 or to the latest 5.4 patch release?

@davidchisnall It merges https://github.com/torvalds/linux/releases/tag/v5.4, which is the original 5.4.0 release (from November 2019). Once we are happy with this, we should do a separate PR to bump us to 5.4.61 (https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tag/?h=v5.4.61).

File: fs/readdir.c (hunk 1 above)
Comment (prp) [reply fs/readdir.c]: No, it doesn't return success. It returns failure, which then makes this code fail, as it apparently relies on `dirent` being sometimes NULL.

Comment (davidchisnall) [reply fs/readdir.c]: Is this broken for other platforms?

Comment (prp) [reply fs/readdir.c]: I don't know. As I said here (https://github.com/lsds/sgx-lkl/pull/822#issuecomment-684676133), this particular code has now been rewritten, and the associated comment mentions that apparently only x86 used to care about `user_access_begin()`.

**Review by davidchisnall — APPROVED** [review]

**Comment by prp** [comment]
@bodzhang do you have further issues, or can you approve this and the associated SGX-LKL PR?

File: fs/readdir.c (hunk 1 above)
Comment (bodzhang) [reply fs/readdir.c]: > @bodzhang do you have further issues, or can you approve this and the associated SGX-LKL PR?

I think we should wait for the [user-space separation PR](https://github.com/lsds/sgx-lkl/pull/818) to get merged first.

**Review by bodzhang — APPROVED** [review]
Good to go.

**Pull request merged**